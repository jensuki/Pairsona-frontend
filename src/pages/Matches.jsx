import { useState, useEffect, useContext } from 'react';
import { useNavigate, Link } from 'react-router-dom';
import Api from '../helpers/api';
import UserContext from '../context/UserContext';
import getProfilePic from '../helpers/getProfilePic';
import Loading from '../components/Loading';
import '../styles/Matches.css';

import noMatches from '../assets/placeholders/no_matches.png';

/** component for fetching list of compatible matches
 *
 * if no mbti is set for curr user, links user to quiz
*/

const Matches = () => {
    const { currUser, loading } = useContext(UserContext);
    const navigate = useNavigate();

    const [matches, setMatches] = useState(null);
    const [error, setError] = useState(null);

    // fetch matches on mount or curr user changes
    useEffect(() => {
        const fetchMatches = async () => {
            if (!currUser?.mbti) {
                setMatches([]); // no mbti no matches
                return;
            }

            try {
                const fetchedMatches = await Api.getMatches(currUser.username);
                setMatches(fetchedMatches);
            } catch (err) {
                setError(err.message)
            }
        };

        fetchMatches();
    }, [currUser]);

    if (error) return <p className="error">{error}</p>
    if (loading || matches == null) {
        return <Loading />;
    }
    // redirect to quiz if no mbti set
    if (!currUser || !currUser.mbti) {
        return (
            <div className="matches-prompt" data-aos="fade-up">
                <h1>Have you taken the quiz?</h1>
                <p>Discover your MBTI personality type by completing a quick 30-question quiz. Start now and gain insights into your unique personality traits</p>
                <Link to="/quiz" className="quiz-btn">
                    Take the Quiz
                </Link>
            </div>
        );
    }

    // handle clicking on matched users profile
    const handleViewProfile = (username) => {
        navigate(`/profile/${username}`);
    };

    return (
        <div className="Matches">
            <h2>Compatible Matches</h2>
            {matches.length > 0 ? (
                <div className="cards" data-aos="fade-up">
                    {matches.map((user) => (
                        <div key={user.username} className="card">
                            {/* matched users profile card */}
                            <h3 className="user-info">
                                <span>{user.firstName} {user.lastName}</span>
                                <span className="mbti">{user.mbti}</span>
                            </h3>
                            <hr></hr>
                            <div className="view-profile">
                                <img
                                    src={getProfilePic(user.profilePic)}
                                    alt={`${user.firstName} ${user.lastName}`}
                                    className="profile-pic"
                                    loading="lazy"
                                />
                                <button onClick={() => handleViewProfile(user.username)}>View Profile</button>
                            </div>
                        </div>
                    ))}
                </div>
            ) : (
                // if no matches
                <img className="no-matches" src={noMatches} alt="No Matches" />
            )}
        </div>
    );
};

export default Matches;
