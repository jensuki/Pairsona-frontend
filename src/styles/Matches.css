.Matches {
    margin-bottom: 70px;
}

.matches-prompt {
    max-width: var(--max-width-lg);
}

.matches-prompt h1 {
    font-family: 'Jost';
    color: #333;
}

.matches-prompt p {
    font-size: 1.1rem;
}

.quiz-btn {
    display: inline-block;
    background-color: var(--primary-green);
    color: var(--btn-white-text);
    padding: 12px 24px;
    font-size: var(--font-size-base);
    font-weight: var(--font-bold);
    transition: var(--transition-bg);
}

.quiz-btn:hover {
    background-color: var(--hover-green);
}

.Matches h3 {
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin-top: 0;
}

.mbti {
    background: rgb(251, 221, 226);
    padding: 5px;
    border: 1px solid #cfc2c2;
    border-radius: var(--border-radius-sm);
}

.Matches button {
    align-self: first baseline;
}

.view-profile {
    margin-top: 20px;
    display: flex;
    justify-content: space-between;
    align-items: center;
}

.no-matches {
    width: 260px;
    height: 200px;
}