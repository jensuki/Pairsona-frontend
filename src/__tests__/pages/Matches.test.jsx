import { render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from '../../test_common/setupTests';
import Matches from '../../pages/Matches';
import Api from '../../helpers/api';
import { describe, expect, test, vi, beforeEach } from 'vitest';
import { demoUser, demoMatches } from '../../test_common/mockData';

// mock api.js
vi.mock('../../helpers/api', () => ({
    default: {
        getMatches: vi.fn()
    }
}))

beforeEach(() => {
    vi.clearAllMocks()
    // explicity reset to avoid any residual behavior and ensure isolated tests
    Api.getMatches.mockReset();
})

// helper
const renderMatches = (currUser = demoUser) => {
    return render(
        <MemoryRouter>
            <UserProvider currUser={currUser}>
                <Matches />
            </UserProvider>
        </MemoryRouter>
    )
}

describe('Matches component', () => {
    test('renders without crashing', async () => {
        renderMatches();

        // expect loading spinner
        await waitFor(() => {
            expect(screen.getByRole('status')).toBeInTheDocument();
        })
    });

    test('matches snapshot for user with mbti set', async () => {
        Api.getMatches.mockResolvedValueOnce(demoMatches); // mock api to return matches
        const { asFragment } = renderMatches(); // render mock matches for demouser

        // wait for matches to load
        await waitFor(() => {
            expect(screen.getByText(/Compatible Matches/i)).toBeInTheDocument();
        })

        expect(asFragment()).toMatchSnapshot();
    })

    test('matches snapshot for user without mbti set', async () => {
        const userNoMbti = { ...demoUser, mbti: null }; // create mock user without mbti set
        // mock get matches to return empty array since no matches should be found
        Api.getMatches.mockResolvedValue([]);

        const { asFragment } = renderMatches(userNoMbti);

        // wait for quiz prompt to render
        await waitFor(() => {
            expect(screen.getByText(/Have you taken the quiz\?/i)).toBeInTheDocument();
        })

        expect(asFragment()).toMatchSnapshot();
    })

    test('displays compatible matches when available', async () => {
        Api.getMatches.mockResolvedValueOnce(demoMatches);
        renderMatches();

        await waitFor(() => {
            demoMatches.forEach((match) => {
                const fullName = `${match.firstName} ${match.lastName}`;
                expect(screen.getByText(new RegExp(fullName, 'i'))).toBeInTheDocument();
            });
        });
    });

    test('handles api error gracefully', async () => {
        Api.getMatches.mockRejectedValue(new Error('Failed to fetch matches')); // mock error msg

        renderMatches();

        // wait for component to finish rendering
        await waitFor(() => {
            expect(screen.queryByText(/Compatible Matches/i)).not.toBeInTheDocument();
            expect(screen.getByText(/Failed to fetch matches/i)).toBeInTheDocument();
        });
    });
});
