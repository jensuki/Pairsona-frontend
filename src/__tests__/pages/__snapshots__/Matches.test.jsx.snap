// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Matches component > matches snapshot for user with mbti set 1`] = `
<DocumentFragment>
  <div
    class="Matches"
  >
    <h2>
      Compatible Matches
    </h2>
    <div
      class="cards"
      data-aos="fade-up"
    >
      <div
        class="card"
      >
        <h3
          class="user-info"
        >
          <span>
            User One
          </span>
          <span
            class="mbti"
          >
            ENFP
          </span>
        </h3>
        <hr />
        <div
          class="view-profile"
        >
          <img
            alt="User One"
            class="profile-pic"
            loading="lazy"
            src="/assets/default_pic.png"
          />
          <button>
            View Profile
          </button>
        </div>
      </div>
      <div
        class="card"
      >
        <h3
          class="user-info"
        >
          <span>
            User Two
          </span>
          <span
            class="mbti"
          >
            ENTP
          </span>
        </h3>
        <hr />
        <div
          class="view-profile"
        >
          <img
            alt="User Two"
            class="profile-pic"
            loading="lazy"
            src="/assets/default_pic.png"
          />
          <button>
            View Profile
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Matches component > matches snapshot for user without mbti set 1`] = `
<DocumentFragment>
  <div
    class="matches-prompt"
    data-aos="fade-up"
  >
    <h1>
      Have you taken the quiz?
    </h1>
    <p>
      Discover your MBTI personality type by completing a quick 30-question quiz. Start now and gain insights into your unique personality traits
    </p>
    <a
      class="quiz-btn"
      data-discover="true"
      href="/quiz"
    >
      Take the Quiz
    </a>
  </div>
</DocumentFragment>
`;
